### PR TITLE
Fix #8594 add -fprofile-update=atomic to fix negative counts in the gcov profile

### LINKED ISSUE #8594: coverage tests compiled without -fprofile-update=atomic results in corrupted counters
### What happens?

TL;DR the code coverage check needs to compile with the flag -fprofile-update=atomic or the profile gets corrupted and some of the execution counters end up negative which results in false positives (uncovered) reports by lcov.

to validate that counters are actually negative see the script in the 'to reproduce' section.

https://gcc.gnu.org/onlinedocs/gcc/Instrumentation-Options.html
>-fprofile-update=method
>
>    Alter the update method for an application instrumented for profile feedback based optimization. The method argument should be one of ‘single’, ‘atomic’ or ‘prefer-atomic’. The first one is useful for single-threaded applications, while the second one prevents profile corruption by emitting thread-safe code.

the fix is easy -- add the flag. compiling with -fprofile-update=atomic eliminates negative counts in the output which prevents false positive (uncovered) reports.

I have a PR which implements the fix.

### To Reproduce

the script below runs the coverage tests, generates the gcov json reports and then queries the report for the number of negative counters. In this example there were 238 samples with negative counts.

```sh
make coverage-check

# time passes...

mkdir cov
cd cov
find ../build/coverage -name '*.gcda' -print0 | \
        xargs -0 -P 0 -n 1 gcov --json-format -x > /dev/null

duckdb <<SQL
/*
extract the per line execution records nested in the json gcov output.
https://gcc.gnu.org/onlinedocs/gcc/Invoking-Gcov.html
{
  "block_ids": ["$block_id"],
  "branches": ["$branch"],
  "count": 2,
  "line_number": 15,
  "unexecuted_block": false,
  "function_name": "foo",
}
count: number of executions of the line
line_number: line number
...
*/

with
gcov_import as (
  select unnest(files, recursive := true), * exclude files
  from '*.gcov.json.gz'
),
gcov_lines as (
  select unnest(lines, recursive := true), * exclude (lines, functions)
  from gcov_import
)
select
  sum((count < 0)::int) as "negative counts"
from
  gcov_lines
SQL
```
| negative counts |
|-----------------|
| 238             |

### OS:

Ubuntu 22.04.3 LTS

### DuckDB Version:

v0.8.2-dev

### DuckDB Client:

NA

### Full Name:

robert manchester

### Affiliation:

unaffiliated

### Have you tried this on the latest `master` branch?

I have tested with a master build

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] Yes, I have

### PR DESCRIPTION
**Pull request by ghost**
Fixes #8594

This change adds a Coverage build type to cmake that adds the following flags:

- -fprofile-update=atomic to eliminate corruption of profile data
- -fprofile-exclude-files=`cat lcov_exclude`. prevents files matching the pattern in lcov_exclude from being instrumented. This speeds up the test runs.
- -fprofile-abs-path resolve ./ .. making later group by file name queries just work

The check_coverage.sh script is slightly modified: 
- move the build out into the makefile / cmake. This will allow the CI to use ninja for a parallel build.
- stops building python for coverage, it isn't used, but the extensions it added are needed so just pull those out.

Testing: verified coverage runs succeed. validated that the corrupted counters are no longer corrupted -- see below. Validated that excluded files are not present in the profile.

gcov can generate json output which duckdb can easily read and query. For example, to validate there are no negative execution counts, starting at the root of the source tree:

```sh
make coverage-check

# time passes...

mkdir cov
cd cov
find ../build/coverage -name '*.gcda' -print0 | \
        xargs -0 -P 0 -n 1 gcov --json-format -x > /dev/null

duckdb <<SQL
/*
extract the per line execution records nested in the json gcov output.
https://gcc.gnu.org/onlinedocs/gcc/Invoking-Gcov.html
{
  "block_ids": ["$block_id"],
  "branches": ["$branch"],
  "count": 2,
  "line_number": 15,
  "unexecuted_block": false,
  "function_name": "foo",
}
count: number of executions of the line
line_number: line number
...
*/

with
gcov_import as (
  select unnest(files, recursive := true), * exclude files
  from '*.gcov.json.gz'
),
gcov_lines as (
  select unnest(lines, recursive := true), * exclude (lines, functions)
  from gcov_import
)
select
  sum((count < 0)::int) as "negative counts"
from
  gcov_lines
SQL
```

### CODE REVIEW

**Comment by robert-manchester** [comment]
`-fprofile-abs-path` is a nice to have if you happen to be querying the gcov output and want to aggregate the counters based on file name. there are a handful of headers that get included using paths relative to the build dir.

FWIW: this change results in a 1.3x speedup (18233 -> 13625 sec) of the unittest runtime (not including the compile or lcov steps).

If the following 3 files are added to lcov_exclude and thus not instrumented the speedup is 2.5x (18233 -> 7271 sec)
```
*/duckdb/common/helper.hpp
*/duckdb/common/unique_ptr.hpp
*/types/selection_vector.hpp
```

i have some changes which runs the unittests in parallel https://github.com/robert-manchester/duckdb/tree/parallel_cov_lcov_2 and then the speedup is roughly 14x (17997 -> 1292 sec) when above files are not instrumented, or 7.7x (17997 -> 2323 sec) when they are instrumented. 17997 sec is used rather than 18233 for the baseline as there are 3 tests which fail consistently on a loaded system which are skipped for parallel runs. The speedup is around 5.5x when comparing the parallel versions to this change.

all numbers collected on a 6 core / 12 thread amd 2600 cpu.
numbers presented are the median of at least 3 runs.

**Comment by Mytherin** [comment]
Thanks for the PR and all the work! Nice find on the atomic counters.

> -fprofile-abs-path is a nice to have if you happen to be querying the gcov output and want to aggregate the counters based on file name. there are a handful of headers that get included using paths relative to the build dir.

The reason I removed it is that it does not seem to be supported on MacOS, which is what we all use for development. The rest of the options seem to work.

> FWIW: this change results in a 1.3x speedup (18233 -> 13625 sec) of the unittest runtime (not including the compile or lcov steps).

Hm, somehow it seems that the CI time has gone up significantly ([4h47m](https://github.com/duckdb/duckdb/actions/runs/6015142603) for this PR, vs [2h8m](https://github.com/duckdb/duckdb/actions/runs/5920509686/job/16051927255) for main) and the CI seems to fail as well. Could you perhaps have a look?

> If the following 3 files are added to lcov_exclude and thus not instrumented the speedup is 2.5x (18233 -> 7271 sec)

That would be great as well - we can definitely skip codecov on a few common files if the speedup is that large.

> i have some changes which runs the unittests in parallel https://github.com/robert-manchester/duckdb/tree/parallel_cov_lcov_2 and then the speedup is roughly 14x (17997 -> 1292 sec) when above files are not instrumented, or 7.7x (17997 -> 2323 sec) when they are instrumented. 17997 sec is used rather than 18233 for the baseline as there are 3 tests which fail consistently on a loaded system which are skipped for parallel runs. The speedup is around 5.5x when comparing the parallel versions to this change.

That is definitely be interesting for local development.

Thanks for all the investigation!

**Comment by robert-manchester** [comment]
> > FWIW: this change results in a 1.3x speedup (18233 -> 13625 sec) of the unittest runtime (not including the compile or lcov steps).
> 
> Hm, somehow it seems that the CI time has gone up significantly ([4h47m](https://github.com/duckdb/duckdb/actions/runs/6015142603) for this PR, vs [2h8m](https://github.com/duckdb/duckdb/actions/runs/5920509686/job/16051927255) for main) 

oof -- that's bad. i was just comparing builds with various patterns used for `-fprofile-exclude-files=pattern` but always with the `-fprofile-update=atomic`.

locally, i can get the runtime close to main by adding the top 3 files from the below table to the exclude pattern.  That gives a 0.97 speedup from main (7193 -> 7384 sec) so only a couple of minutes slower. However, when using the github action from my fork main is around 2.5 hours where this change (or variants of with extra exclude patterns) hovers around 3-4 hours. Parallelizing the unittests brings the github action time to 2-2.5 hours but that change is 500 lines of python/makefile and adds a bit of complexity.

dissembling the build we can see that updating the counters without the atomic flag the counter increment is 3 instructions.
```asm
mov    0xe314067(%rip),%rax # <__gcov0._XXNormalizeIntervalEXX>
add    $0x1,%rax
mov    %rax,0xe31405c(%rip) # <__gcov0._XXNormalizeIntervalEXX>
```
and when using the atomic flag replaced with a slower atomic increment instruction.
```asm
lock addq $0x1,0xe314067(%rip) # <__gcov0._XXNormalizeIntervalEXX>
```
 
however, we really do not care if counts go missing -- the racy version is just fine for showing that the code executed as there will never be a transition to zero from some positive number in the above asm. But the problem is with corrupted counters being negative, which must happen somewhere else, and lcov treating negative counts as zero which can trigger false positive (uncovered failures).


| rank |  sum(count)  | pct  | tot pct |                       file                       |
|------|--------------|------|---------|--------------------------------------------------|
| 1    | 343144801202 | 45.0 | 45.0    | include/duckdb/common/helper.hpp                 |
| 2    | 79181675437  | 10.0 | 55.0    | include/duckdb/common/unique_ptr.hpp             |
| 3    | 34456544118  | 4.0  | 59.0    | common/assert.cpp                                |
| 4    | 28105544075  | 4.0  | 63.0    | include/duckdb/common/types/validity_mask.hpp    |
| 5    | 20362843686  | 3.0  | 66.0    | include/duckdb/common/vector.hpp                 |
| 6    | 17069213269  | 2.0  | 68.0    | include/duckdb/common/types/selection_vector.hpp |
| 7    | 15900568012  | 2.0  | 70.0    | include/duckdb/common/types/string_type.hpp      |
| 8    | 12027966323  | 2.0  | 72.0    | common/types/selection_vector.cpp                |
| 9    | 11558298126  | 2.0  | 74.0    | common/types.cpp                                 |
| 10   | 9068690856   | 1.0  | 75.0    | include/duckdb/execution/index/art/node.hpp      |


>
> and the CI seems to fail as well.

so that failure is somewhat interesting. my fork runs the CodeCoverage check as well whenever a change is pushed and it was [fine](https://github.com/robert-manchester/duckdb/actions/runs/6015137821). I looked into the failure on the duckdb CI and the files and there was one interesting [failure](https://github.com/duckdb/duckdb/actions/runs/6015142603/job/16316510966).
`include/duckdb/main/db_instance_cache.hpp:19:DBInstanceCache() {};` is only hit by one test (i ran every test individually and queried the gcov data -- duckdb is pretty handy).
|                  test                   | sum(count) | list(line_number) |
|-----------------------------------------|------------|-------------------|
| Issue #6931 - test windows unicode path | 3          | [19]              |

so looking over the logs of the failed CI run it did run the above test so why the counts ended up at zero is somewhat of a mystery as the test looks deterministic -- don't think there is even an if statement.

without the raw gcov data i can only guess but I expect that there are similar failures which pop up from time to time and are unrelated to this change. I'd like to collect the gcov data and stuff it in the zip file along with the html to have a little more info when failures occur. I was running the local parallel version overnight and was unable to reproduce.

**Comment by robert-manchester** [comment]
> and the CI seems to fail as well.

TL;DR seems that the `ub_test_api.cpp.gcda` counters did not get written out fully / correctly.

looking a little more at the undercovered files from the failure [failure](https://github.com/duckdb/duckdb/actions/runs/6015142603/job/16316510966). it seems that they all should be covered in one object file, `ub_test_api.cpp.gcda`. When querying the gcov data from a test run for lines that only appear in  `ub_test_api.cpp.gcda` we get this list.

|         line numbers       | lines |                    file                     |
|----------------------------|-------|---------------------------------------------|
| [101, 102, 103]            | 3     | include/duckdb/main/appender.hpp            |
| [81, 82, 83, 232, 233, 234]| 6     | include/duckdb/main/connection.hpp          |
| [19]                       | 1     | include/duckdb/main/db_instance_cache.hpp   |
| [91, 92, 93, 167, ..., 174]| 8     | include/duckdb/main/prepared_statement.hpp  | 
| [131, 132, 138, ..., 196]  | 32    | include/duckdb/main/query_result.hpp        | 
| [88, 89]                   | 2     | include/duckdb/planner/logical_operator.hpp |
| [24, 25, 32, 33, ..., 75]  | 17    | include/duckdb/storage/object_cache.hpp     |

which is about what is missing in the lcov output -- except for connection.hpp [232,233,234] which are covered. appender.hpp and connection.hpp do not appear in the error log because they have enough buffer in the uncovered_files.csv.

so `ub_test_api.cpp.gcda` did not get written out correctly. the docs say that all threads must be joined for the gcda (counter files) to be properly written. maybe the thread was not joined. collecting the gcov output in the zip file will assist with tracking down errors like this in the future.

**Comment by robert-manchester** [comment]
excluding the top 8 files with the most counts got the github action time down to [2:40](https://github.com/robert-manchester/duckdb/actions/runs/6042976253/job/16399237778) (sample size = 1).

**Comment by robert-manchester** [comment]
> dissembling the build we can see that updating the counters without the atomic flag the counter increment is 3 instructions.
> 
> ```assembly
> mov    0xe314067(%rip),%rax # <__gcov0._XXNormalizeIntervalEXX>
> add    $0x1,%rax
> mov    %rax,0xe31405c(%rip) # <__gcov0._XXNormalizeIntervalEXX>
> ```
> 
> and when using the atomic flag replaced with a slower atomic increment instruction.
> 
> ```assembly
> lock addq $0x1,0xe314067(%rip) # <__gcov0._XXNormalizeIntervalEXX>
> ```
> 
> however, we really do not care if counts go missing -- the racy version is just fine for showing that the code executed as there will never be a transition to zero from some positive number in the above asm. But the problem is with corrupted counters being negative, which must happen somewhere else.

so the negative counts are a result of the counters being out of sync. consider the following:

```
1: void myfunc(int arg) {
2:    if (arg > 0) {
3:      return  myfunc(arg - 1);
4:    }
5:  return 0;
6: }
```
this will result in 2 counters emitted. a counter incremented on function entry and then another incremented when the branch is taken. The branch not taken can be calculated by `func_count - branch_count`, thus when the function is accessed concurrently and the counters are not atomic the count computed for line 5 can be inconsistent and misrepresent if the code was executed or not.

**Comment by robert-manchester** [comment]
FWIW: the coverage build has slowed down a bit over the past week or two. merge [625a209](https://github.com/duckdb/duckdb/pull/8595/commits/625a2093e7d127304d4fdfa7376ba27807201e0f) put the runtime back in the 3-4 hour range on github action (sample size of 2)

**Comment by robert-manchester** [comment]
closed due to inability to test change.
